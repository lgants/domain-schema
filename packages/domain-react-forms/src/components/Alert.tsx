import * as React from 'react';
import { Alert as RSAlert } from 'reactstrap';

export interface Props {
  children?: any;
  color?: string;
}

const Alert = ({ children, color, ...props }: Props) => {
  if (color === 'error') {
    color = 'danger';
  }
  return (
    <RSAlert {...props} color={color}>
      {children}
    </RSAlert>
  );
};

export default Alert;
