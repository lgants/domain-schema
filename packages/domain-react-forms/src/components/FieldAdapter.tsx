import * as PropTypes from 'prop-types';
import * as React from 'react';
import { Component } from 'react';

export interface Props {
  name: string;
  component?: any;
  onChange?: any;
  value?: string;
  defaultValue?: string;
  checked?: boolean;
  defaultChecked?: boolean;
  disabled?: boolean;
}

export default class Field extends Component {
  public static contextTypes = {
    formik: PropTypes.object
  };

  constructor(public props: Props, public context: any) {
    super(props, context);
    if (!context.formik) {
      throw new Error('Field must be inside a component decorated with formik()');
    }
  }

  public render() {
    const { formik } = this.context;
    const { component, name, value, defaultValue, checked, defaultChecked, onChange, disabled } = this.props;

    const meta = {
      touched: formik.touched[name],
      error: formik.errors[name]
    };

    const input = {
      onChange,
      onBlur: formik.handleBlur,
      name,
      value,
      checked,
      defaultValue,
      defaultChecked,
      disabled
    };

    return React.createElement(component, {
      ...this.props,
      input,
      meta
    });
  }
}
