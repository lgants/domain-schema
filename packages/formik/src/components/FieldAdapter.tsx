import * as PropTypes from 'prop-types';
import * as React from 'react';
import { Component } from 'react';

export interface Props {
  attrs: any;
  checked?: boolean;
  component: any;
  fieldType?: string;
  name: string;
  onBlur?: any;
  onChange?: any;
  options?: any;
  parent?: any;
  value?: string | number;
}

export default class Field extends Component<Props, {}> {
  public static contextTypes = {
    formik: PropTypes.object
  };

  constructor(public props: Props, public context: any) {
    super(props, context);
  }

  public render() {
    const { formik: { setFieldValue, handleChange, handleBlur, touched, errors } } = this.context;
    const { component, parent, attrs, fieldType, name } = this.props;
    const { onChange, onBlur, type } = attrs;

    const input = {
      ...attrs,
      onChange:
        parent && parent.name
          ? e =>
              setFieldValue(parent.name, {
                ...parent.value,
                [name]: e.target.value
              })
          : onChange ? onChange : handleChange,
      onBlur: onBlur ? onBlur : handleBlur,
      type: type || (fieldType === 'input' ? 'text' : fieldType),
      name
    };

    const meta = {
      touched: touched[name],
      error: errors[name]
    };

    return React.createElement(component, {
      ...this.props,
      input,
      meta
    });
  }
}
