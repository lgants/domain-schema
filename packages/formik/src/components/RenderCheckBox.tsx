import * as React from 'react';
import { FormFeedback, FormGroup, Input, Label } from 'reactstrap';

import { RenderComponentProps } from '../types';

const RenderCheckBox = ({ input, options, meta: { touched, error } }: RenderComponentProps) => {
  return (
    <FormGroup check {...options}>
      <Label check>
        <Input {...input} /> {input.label}
        {touched && (error && <div className="validation-err">{error}</div>)}
      </Label>
    </FormGroup>
  );
};

export default RenderCheckBox;
