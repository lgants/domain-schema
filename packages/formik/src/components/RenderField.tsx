import * as React from 'react';
import { FormFeedback, FormGroup, Input, Label } from 'reactstrap';

import { RenderComponentProps } from '../types';

const RenderField = ({ input, options, meta: { touched, error }, children }: RenderComponentProps) => {
  const { label, placeholder } = input;
  let valid: boolean = null;
  if (touched && error) {
    valid = false;
  }

  return (
    <FormGroup {...options}>
      {label && <Label>{label}</Label>}
      <Input {...input} placeholder={placeholder || label || ''} valid={valid}>
        {children}
      </Input>
      {touched && (error && <FormFeedback>{error}</FormFeedback>)}
    </FormGroup>
  );
};

export default RenderField;
