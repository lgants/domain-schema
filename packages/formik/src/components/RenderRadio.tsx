import * as React from 'react';
import { FormFeedback, FormGroup, Input, Label } from 'reactstrap';

import { RenderComponentProps } from '../types';

const RenderRadio = ({ input, options, meta: { touched, error } }: RenderComponentProps) => {
  return (
    <FormGroup tag="fieldset" {...options} value="Term 1 ...">
      {input.label && <legend>{input.label}</legend>}
      {input.values.map(radio => {
        return radio.value ? (
          <FormGroup key={radio.value} check>
            <Label check>
              <Input {...input} value={radio.value} checked={radio.value === input.value} /> {radio.label}
            </Label>
          </FormGroup>
        ) : (
          <FormGroup key={radio} check>
            <Label check>
              <Input {...input} value={radio} checked={radio === input.value} /> {radio}
            </Label>
          </FormGroup>
        );
      })}
      {touched && (error && <div className="validation-err">{error}</div>)}
    </FormGroup>
  );
};

export default RenderRadio;
