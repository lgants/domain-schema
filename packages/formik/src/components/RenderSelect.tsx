import * as React from 'react';
import { FormFeedback, FormGroup, Input, Label } from 'reactstrap';

import { RenderComponentProps } from '../types';
import Option from './Option';

const RenderSelect = ({ input, options, meta: { touched, error } }: RenderComponentProps) => {
  const { label, values } = input;
  let className = input.className || '';
  if (touched && error) {
    className += ' invalid-select';
  }

  return (
    <FormGroup {...options}>
      {label && <Label>{label}</Label>}
      <Input {...input} className={className}>
        {values.map(option => {
          return option.value ? (
            <Option key={option.value} value={option.value}>
              {option.label}
            </Option>
          ) : (
            <Option key={option} value={option}>
              {option}
            </Option>
          );
        })}
      </Input>
      {touched && (error && <div className="validation-err">{error}</div>)}
    </FormGroup>
  );
};

export default RenderSelect;
