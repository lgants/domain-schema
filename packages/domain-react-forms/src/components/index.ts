import Alert from './Alert';
import Button from './Button';
import Field from './FieldAdapter';
import Form from './Form';
import Option from './Option';
import RenderCheckBox from './RenderCheckBox';
import RenderField from './RenderField';
import RenderRadio from './RenderRadio';
import RenderSelect from './RenderSelect';

export { Alert, Button, Field, Form, Option, RenderRadio, RenderCheckBox, RenderField, RenderSelect };
