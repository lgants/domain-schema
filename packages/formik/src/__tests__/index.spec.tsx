import DomainSchema, { Schema } from '@domain-schema/core';
import { FieldValidators } from '@domain-schema/validation';
import * as React from 'react';
import renderer from 'react-test-renderer';

import { DomainSchemaFormik, FieldTypes, FormSchema, FSF } from '../';
import { Button, Form, RenderCheckBox, RenderField, RenderSelect } from '../components';

DomainSchemaFormik.setFormComponents({
  input: RenderField,
  checkbox: RenderCheckBox,
  select: RenderSelect,
  button: Button,
  form: Form
});

describe('DomainFormik', () => {
  it('should generate simple form', () => {
    const schema = new DomainSchema(
      class extends FormSchema {
        public __ = { name: 'User' };
        public username: FSF = {
          type: String,
          fieldType: DomainSchemaFormik.fields.input,
          input: {
            label: 'Username',
            placeholder: 'name'
          },
          defaultValue: 'John'
        };
        public email: FSF = {
          type: String,
          fieldType: DomainSchemaFormik.fields.input,
          input: {
            type: 'email',
            label: 'Email'
          }
        };
        public pass: FSF = {
          type: String,
          fieldType: DomainSchemaFormik.fields.input,
          input: {
            type: 'password',
            label: 'Email'
          }
        };
        public isActive: FSF = {
          type: String,
          fieldType: DomainSchemaFormik.fields.checkbox,
          input: {
            label: 'Is Active'
          },
          defaultValue: true
        };
      }
    );
    const form = new DomainSchemaFormik(schema);
    const FormComponent = form.generateForm();
    const component = renderer.create(<FormComponent onSubmit={() => null} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should generate complex form', () => {
    class Profile extends Schema {
      public __ = { name: 'Profile' };
      public firstName: FSF = {
        type: String,
        fieldType: DomainSchemaFormik.fields.input,
        input: {
          label: 'First Name'
        },
        defaultValue: 'Ashley',
        required: {
          value: true,
          msg: 'Required First Name'
        }
      };
      public lastName: FSF = {
        type: String,
        fieldType: DomainSchemaFormik.fields.input,
        input: {
          label: 'Last Name'
        },
        required: true
      };
    }
    const schema = new DomainSchema(
      class User extends FormSchema {
        public __ = { name: 'User' };
        public username: FSF = {
          type: String,
          fieldType: DomainSchemaFormik.fields.input,
          input: {
            type: 'text',
            label: 'Username'
          },
          defaultValue: 'User',
          required: true,
          validators: [
            value => {
              return value.length > 3 ? undefined : 'Must Be more than 3 characters';
            }
          ]
        };
        public email: FSF = {
          type: String,
          fieldType: DomainSchemaFormik.fields.input,
          input: {
            type: 'email',
            label: 'Email',
            placeholder: 'User email'
          },
          required: true,
          email: true
        };
        public profile = {
          type: Profile
        };
        public password: FSF = {
          type: String,
          fieldType: DomainSchemaFormik.fields.input,
          input: {
            type: 'password',
            label: 'Password'
          },
          required: true,
          min: 5
        };
        public passwordConfirmation: FSF = {
          type: String,
          fieldType: DomainSchemaFormik.fields.input,
          input: {
            type: 'password',
            label: 'Password Confirmation'
          },
          matches: 'password'
        };
        public setSubmitBtn(): any {
          return {
            label: 'Submit',
            autovalidate: true
          };
        }
      }
    );
    const userForm = new DomainSchemaFormik(schema);
    const FormComponent = userForm.generateForm();
    const component = renderer.create(<FormComponent onSubmit={() => null} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should generate form with custom field', () => {
    const MyComponent = ({ input: { name, label, value, className, ...attrs } }) => {
      return (
        <div>
          <input {...attrs} type="text" name={name} placeholder={label} value={value} className={className} />
        </div>
      );
    };
    const schema = new DomainSchema(
      class User extends FormSchema {
        public __ = { name: 'User' };
        public username: FSF = {
          type: String,
          fieldType: DomainSchemaFormik.fields.input,
          input: {
            inputType: 'text',
            label: 'Username'
          },
          defaultValue: 'User'
        };
        public userfield: FSF = {
          type: String,
          fieldType: DomainSchemaFormik.fields.custom,
          component: MyComponent,
          input: {
            label: 'User Field',
            className: 'my-awesome-field'
          },
          defaultValue: 'my field'
        };
        public setSubmitBtn(): any {
          return {
            label: 'Submit'
          };
        }
      }
    );
    const userForm = new DomainSchemaFormik(schema);
    const FormComponent = userForm.generateForm();
    const component = renderer.create(<FormComponent onSubmit={() => null} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
