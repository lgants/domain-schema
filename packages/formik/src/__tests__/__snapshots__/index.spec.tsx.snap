// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainFormik should generate complex form 1`] = `
<form
  className=""
  name="User"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor={undefined}
    >
      Username
    </label>
    <input
      className="form-control"
      label="Username"
      name="username"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value="User"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor={undefined}
    >
      Email
    </label>
    <input
      className="form-control"
      label="Email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="User email"
      type="email"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor={undefined}
    >
      First Name
    </label>
    <input
      className="form-control"
      label="First Name"
      name="firstName"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="First Name"
      type="text"
      value="Ashley"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor={undefined}
    >
      Last Name
    </label>
    <input
      className="form-control"
      label="Last Name"
      name="lastName"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Last Name"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor={undefined}
    >
      Password
    </label>
    <input
      className="form-control"
      label="Password"
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor={undefined}
    >
      Password Confirmation
    </label>
    <input
      className="form-control"
      label="Password Confirmation"
      name="passwordConfirmation"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Password Confirmation"
      type="password"
      value=""
    />
  </div>
  <div>
    <button
      className="btn btn-secondary disabled"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`DomainFormik should generate form with custom field 1`] = `
<form
  className=""
  name="User"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor={undefined}
    >
      Username
    </label>
    <input
      className="form-control"
      inputType="text"
      label="Username"
      name="username"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value="User"
    />
  </div>
  <div>
    <input
      className="my-awesome-field"
      name="userfield"
      placeholder="User Field"
      value="my field"
    />
  </div>
  <div>
    <button
      className="btn btn-secondary"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`DomainFormik should generate simple form 1`] = `
<form
  className=""
  name="User"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor={undefined}
    >
      Username
    </label>
    <input
      className="form-control"
      label="Username"
      name="username"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="name"
      type="text"
      value="John"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor={undefined}
    >
      Email
    </label>
    <input
      className="form-control"
      label="Email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor={undefined}
    >
      Email
    </label>
    <input
      className="form-control"
      label="Email"
      name="pass"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Email"
      type="password"
      value=""
    />
  </div>
  <div
    className="form-check"
  >
    <label
      className="form-check-label "
      htmlFor={undefined}
    >
      <input
        checked={true}
        className="form-check-input"
        label="Is Active"
        name="isActive"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
      />
       
      Is Active
    </label>
  </div>
  <div>
    <button
      className="btn btn-secondary"
      onClick={[Function]}
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;
